server/trillian_log_server.go

```diff
@@ -56,7 +56,30 @@ func (t *TrillianLogServer) getStorageForLog(logId int64) (storage.LogStorage, e
 
 // QueueLeaves submits a batch of leaves to the log for later integration into the underlying tree.
 func (t *TrillianLogServer) QueueLeaves(ctx context.Context, req *trillian.QueueLeavesRequest) (*trillian.QueueLeavesResponse, error) {
-	return nil, ErrNotImplemented
+	leaves := protosToLeaves(req.Leaves)
+
+	if len(leaves) == 0 {
+		return &trillian.QueueLeavesResponse{Status: buildStatusWithDesc(trillian.TrillianApiStatusCode_ERROR, "Must queue at least one leaf")}, nil
+	}
+
+	tx, err := t.prepareStorageTx(*req.LogId)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = tx.QueueLeaves(leaves)
+
+	if err != nil {
+		tx.Rollback()
```

Comment:
I think I'd like to defer this until the best approach is clearer. I don't like the explicit rollbacks either but the alternative needs thought.